File: competition/node/bench/bench.js
// Node.js (V8) translation of competition/bench/workloads.scm -- same nine
// workloads, timed the same way, for a same-machine cross-language
// comparison alongside the creme/Crystal/Go/Ruby/Racket/Guile numbers in
// competition/bench.scm.
//
// hashtable-test exercises V8's own real native Map under a mixed read/
// write/growth workload -- see that function's own comment for the full
// design (shared identically, in shape, across every language's own
// bench.* port). record-test uses a distinct Point
// per index -- see that function's own comment for why a single reused
// instance doesn't give a meaningful measurement here (V8's own JIT can
// hoist/eliminate a provably-invariant reused object just as an AOT
// compiler can).
//
// One workload is translated idiomatically rather than literally, since V8
// has no guaranteed tail-call optimization (mirroring bench.rb's own note
// for MRI):
//   - sum-to: an iterative loop (a literal recursive translation would blow
//     the stack at n=2,000,000 under V8, which doesn't TCO by default).
// build-list uses a genuine singly-linked Cons class (mirroring bench.cr's
// own Cons), consing onto the front in O(1) and reversing/measuring via
// manual pointer-chasing -- not a native Array's push/pop, so this stays a
// fair comparison for exactly the allocation-pressure workload it's meant
// to be.
// vector-sum-test uses a plain Array as the "vector" (a fixed-size,
// index-set/get generic array, not a typed array like Float64Array, to
// match the other variants' generic-vector semantics).
// string-build-test uses `+=` (V8 ropes/flattens repeated string
// concatenation efficiently, matching the Scheme benchmark's own
// string-output-port builder idiom).

function fib(n) {
  return n < 2 ? n : fib(n - 1) + fib(n - 2);
}

function sumTo(n, acc) {
  while (n > 0) {
    acc += n;
    n -= 1;
  }
  return acc;
}

class Cons {
  constructor(car, cdr) {
    this.car = car;
    this.cdr = cdr;
  }
}

function buildList(n) {
  let acc = null;
  for (let i = 0; i < n; i += 1) {
    acc = new Cons(i, acc);
  }
  return acc;
}

function listReverse(list) {
  let acc = null;
  let node = list;
  while (node) {
    acc = new Cons(node.car, acc);
    node = node.cdr;
  }
  return acc;
}

function listLength(list) {
  let n = 0;
  let node = list;
  while (node) {
    n += 1;
    node = node.cdr;
  }
  return n;
}

function vectorSumTest(n) {
  const v = new Array(n).fill(0);
  for (let i = 0; i < n; i += 1) {
    v[i] = i * 2;
  }
  let acc = 0;
  for (let i = 0; i < n; i += 1) {
    acc += v[i];
  }
  return acc;
}

// hashtable-test against V8's own real, native Map (not a plain Object --
// Map is the genuine hash-table type here, with real key hashing and no
// legacy dictionary-mode/prototype-chain baggage) -- see competition/
// scheme/bench/creme.scm's own comment for the full mixed read/write/
// growth design and why this workload is defined per-language rather
// than a hand-rolled shared algorithm. String keys ("k" + i), not raw
// integers -- matches every other language's own bench.* port so the
// same checksum is comparable across all of them.
function hashtableTest(n) {
  const h = new Map();
  for (let i = 0; i < n; i += 1) {
    h.set(`k${i}`, i * 2);
  }
  let acc = 0;
  for (let i = 0; i < n; i += 1) {
    switch (i % 4) {
      case 0:
        h.set(`k${n + i}`, i);
        break;
      case 1: {
        const key = `k${i % n}`;
        h.set(key, h.get(key) + 1);
        break;
      }
      default:
        acc += h.get(`k${i % n}`);
    }
  }
  return acc;
}

class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

// A distinct Point per index (like vectorSumTest's distinct number per
// index), not one reused mutable instance -- see workloads.scm's own
// comment on record-test for why that shape isn't a meaningful benchmark
// under a sufficiently smart JIT/AOT compiler.
function recordTest(n) {
  const v = new Array(n);
  for (let i = 0; i < n; i += 1) {
    v[i] = new Point(i, i * 2);
  }
  let acc = 0;
  for (let i = 0; i < n; i += 1) {
    const p = v[i];
    acc += p.x + p.y;
  }
  return acc;
}

function stringBuildTest(n) {
  let s = "";
  for (let i = 0; i < n; i += 1) {
    s += "x";
  }
  return s.length;
}

function tak(x, y, z) {
  return y < x ? tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y)) : z;
}

// Positions reuse the same Cons list buildList uses, so nqueens conses a
// position per placement exactly as the Scheme version does.
function queensSafe(col, positions) {
  let node = positions;
  let dist = 1;
  while (node) {
    if (node.car === col || Math.abs(node.car - col) === dist) return false;
    node = node.cdr;
    dist += 1;
  }
  return true;
}

function nqueens(boardSize, row = 0, positions = null) {
  if (row === boardSize) return 1;
  let count = 0;
  for (let col = 0; col < boardSize; col += 1) {
    if (queensSafe(col, positions)) {
      count += nqueens(boardSize, row + 1, new Cons(col, positions));
    }
  }
  return count;
}

function timedRun(name, fn) {
  const start = process.hrtime.bigint();
  const result = fn();
  const elapsed = Number(process.hrtime.bigint() - start) / 1e9;
  console.log(`${name} = ${result}  (${elapsed}s)`);
  return result;
}

const totalStart = process.hrtime.bigint();

timedRun("fib(27)", () => fib(27));
timedRun("sum-to(2000000)", () => sumTo(2000000, 0));
timedRun("build-list(200000) length+reverse", () => listLength(listReverse(buildList(200000))));
timedRun("vector-sum-test(500000)", () => vectorSumTest(500000));
timedRun("hashtable-test(200000)", () => hashtableTest(200000));
timedRun("record-test(500000)", () => recordTest(500000));
timedRun("string-build-test(4000) length", () => stringBuildTest(4000));
timedRun("tak(18,12,6)", () => tak(18, 12, 6));
timedRun("nqueens(9)", () => nqueens(9));

console.log(`total = ${Number(process.hrtime.bigint() - totalStart) / 1e9}s`);

File: competition/ruby/bench/bench.rb
# Ruby translation of competition/bench/workloads.scm — same nine
# workloads, timed the same way, for a same-machine cross-language
# comparison alongside the creme/Racket numbers in competition/bench.scm.
#
# hashtable-test exercises Ruby's own real Hash under a mixed read/write/
# growth workload -- see that method's own comment for the full design
# (shared identically, in shape, across every language's own bench.*
# port). record-test uses a distinct Point
# instance per index -- see that method's own comment for why a single
# reused instance doesn't give a meaningful measurement here.
#
# One workload is translated idiomatically rather than literally, since
# Ruby (MRI) has no guaranteed tail-call optimization:
#   - sum-to: an iterative loop (a literal recursive translation would blow
#     the stack at n=2,000,000 under MRI, which doesn't TCO by default).
# build-list uses a genuine singly-linked Cons class (mirroring bench.cr's
# own Cons), consing onto the front in O(1) and reversing/measuring via
# manual pointer-chasing — same allocate-a-heap-object-per-element pattern
# Scheme's build-list has, not Array#<<'s amortized-O(1)-but-no-per-element-
# allocation behavior, so this stays a fair comparison for exactly the
# allocation-pressure workload it's meant to be.
# vector-sum-test uses a plain Array as the "vector" (Ruby's Array supports
# O(1) amortized index get/set, matching vector-ref/vector-set!).
# string-build-test uses `<<` (Ruby's own mutating, amortized-O(1)-append
# String builder idiom), matching the Scheme benchmark's own switch to a
# string output port — the comparison is each language's best-practice
# string-building approach, not a deliberately-naive allocate-per-append one.

def fib(n)
  n < 2 ? n : fib(n - 1) + fib(n - 2)
end

def sum_to(n, acc)
  while n > 0
    acc += n
    n -= 1
  end
  acc
end

class Cons
  attr_reader :car, :cdr

  def initialize(car, cdr)
    @car = car
    @cdr = cdr
  end
end

def build_list(n)
  acc = nil
  i = 0
  while i < n
    acc = Cons.new(i, acc)
    i += 1
  end
  acc
end

def list_reverse(list)
  acc = nil
  node = list
  while node
    acc = Cons.new(node.car, acc)
    node = node.cdr
  end
  acc
end

def list_length(list)
  n = 0
  node = list
  while node
    n += 1
    node = node.cdr
  end
  n
end

def vector_sum_test(n)
  v = Array.new(n, 0)
  i = 0
  while i < n
    v[i] = i * 2
    i += 1
  end
  acc = 0
  i = 0
  while i < n
    acc += v[i]
    i += 1
  end
  acc
end

# hashtable-test against Ruby's own real Hash (see competition/scheme/
# bench/creme.scm's own comment for the full mixed read/write/growth
# design and why this workload is defined per-language rather than a
# hand-rolled shared algorithm). String keys ("k" + i), not raw integers
# -- matches every other language's own bench.* port so the same
# checksum is comparable across all of them (also happens to matter for
# Lua specifically, though not Ruby: see that file's own comment).
def hashtable_test(n)
  h = {}
  i = 0
  while i < n
    h["k#{i}"] = i * 2
    i += 1
  end
  acc = 0
  i = 0
  while i < n
    case i % 4
    when 0
      h["k#{n + i}"] = i
    when 1
      key = "k#{i % n}"
      h[key] += 1
    else
      acc += h["k#{i % n}"]
    end
    i += 1
  end
  acc
end

Point = Struct.new(:x, :y)

# A distinct Point per index (like vector_sum_test's distinct Integer per
# index), not one reused mutable instance -- see workloads.scm's own
# comment on record-test for why that shape risks being optimized away
# under a smarter (AOT/JIT) host than MRI.
def record_test(n)
  v = Array.new(n) { |i| Point.new(i, i * 2) }
  acc = 0
  i = 0
  while i < n
    p = v[i]
    acc += p.x + p.y
    i += 1
  end
  acc
end

def string_build_test(n)
  s = String.new
  i = 0
  while i < n
    s << "x"
    i += 1
  end
  s.length
end

def tak(x, y, z)
  y < x ? tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y)) : z
end

# Positions reuse the same Cons list build_list uses, so nqueens conses a
# position per placement exactly as the Scheme version does.
def queens_safe?(col, positions)
  node = positions
  dist = 1
  while node
    return false if node.car == col || (node.car - col).abs == dist
    node = node.cdr
    dist += 1
  end
  true
end

def nqueens(board_size, row = 0, positions = nil)
  return 1 if row == board_size
  count = 0
  col = 0
  while col < board_size
    count += nqueens(board_size, row + 1, Cons.new(col, positions)) if queens_safe?(col, positions)
    col += 1
  end
  count
end

def timed_run(name)
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  result = yield
  elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
  puts "#{name} = #{result}  (#{elapsed}s)"
  result
end

total_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)

timed_run("fib(27)") { fib(27) }
timed_run("sum-to(2000000)") { sum_to(2_000_000, 0) }
timed_run("build-list(200000) length+reverse") { list_length(list_reverse(build_list(200_000))) }
timed_run("vector-sum-test(500000)") { vector_sum_test(500_000) }
timed_run("hashtable-test(200000)") { hashtable_test(200_000) }
timed_run("record-test(500000)") { record_test(500_000) }
timed_run("string-build-test(4000) length") { string_build_test(4_000) }
timed_run("tak(18,12,6)") { tak(18, 12, 6) }
timed_run("nqueens(9)") { nqueens(9) }

puts "total = #{Process.clock_gettime(Process::CLOCK_MONOTONIC) - total_start}s"
